### ISSUE TITLE
update role allowances for tr, th, td

### ISSUE THREAD

**Issue by scottaohara** [validator/validator#1125]
when used within a `role=table, grid, treegrid` the `td`, `tr`, and `th` elements should not be allowed any explicitly set role, as that would break the a11y mappings of the table, grid or treegrid.  
[td in role=table](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/td-table-test.html)
[td in role=grid](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/td-grid-test.html)
[td in role=treegrid](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/td-treegrid-test.html)

[tr in role=table](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/tr-table-test.html)
[tr in role=grid](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/tr-grid-test.html)
[tr in role=treegrid](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/tr-treegrid-test.html)

[th in role=table](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/th-table-test.html)
[th in role=grid](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/th-grid-test.html)
[th in role=treegrid](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/th-treegrid-test.html)

expectations:
*  if a `<td>`'s ancestor table element has role=table, grid or treegrid - then no allowed role.
*  if a `<th>`'s ancestor table element has role=table, grid or treegrid - then no allowed role.
*  if a `<tr>`'s ancestor table element has role=table, grid or treegrid - then no allowed role.
* if any use of `cell`, `gridcell`, `row`, `rowheader`, `columnheader` are found on these elements - then those should  produce warnings as being unnecessary

if a td, tr, or th are used within a table with no set role, thus it should be an implicit role=table (regardless of what browser heuristics may do) - then no allowed role
[td within a table testcase](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/td-test.html)
[th within a table testcase](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/th-test.html)
[tr within a table testcase](https://thepaciellogroup.github.io/aria-test-builder/dist/testcases/tr-test.html)

If td, tr, or th are used within a table where it's role is overwritten by a role not mentioned above (e.g., role=none, article, etc), then these elements can continue to allow any role.  **the validator already meets this expectation**

**Comment by sideshowbarker** [comment]
> * if a `<td>`'s ancestor table element has role=table, grid or treegrid - then no allowed role.
> 
> * if a `<th>`'s ancestor table element has role=table, grid or treegrid - then no allowed role.
> 
> * if a `<tr>`'s ancestor table element has role=table, grid or treegrid - then no allowed role.
> 
> * if any use of `cell`, `gridcell`, `row`, `rowheader`, `columnheader` are found on these elements - then those should  produce warnings as being unnecessary

Doesn’t the last requirement in that list conflict with the previous three requirements?

That is, aren’t we already requiring that those td/th/tr elements have no allowed role? And if so, then that includes requiring that they not have `cell`, `gridcell`, `row`, `rowheader`, `columnheader` roles. So how would we ever hit the warning case in the list item in the list above? Won’t we always hit the error case instead?

Or if not, then in what case would we reach that last list item?

**Comment by scottaohara** [comment]
here's an unfortunate situation I've come across where the containing table element needed to be suppressed, and then ARIA roles were needed to re-enable the table element mappings for all the elements, as just using role=table on the `tbody` wasn't enough to ensure that the tr, th, and tds were recognized as the necessary mappings.

```html
<table role=presentation>
	<!-- content that shouldn't be exposed as part of the table -->
	
	<!-- re-enable table semantics -->
	<tbody role=table>
		<tr role=row>
			<th role=columnheader>header 1</th>
			<th role=columnheader>header 2</th>
		</tr>
		<tr role=row>
			<td role=cell>cell 1</td>
			<td role=cell>cell 2</td>
		</tr>
	</tbody>
</table>
```

**Comment by sideshowbarker** [comment]
> yes, i believe you to be correct. the last bullet is unnecessary and goes beyond the normative rule, so no worries there.

OK, thanks for confirming

**Comment by scottaohara** [comment]
Hi @sideshowbarker 

Related to our [above discussion](https://github.com/validator/validator/issues/1125#issuecomment-867818842) 

the following HTML, when run through the validator produces both warnings and errors due to the explicit roles that are being specified:

```html
<table role=table>
  <tr role=row>
    <th role=columnheader>...</th>
    <th role=columnheader>...</th>
  </tr>
  <tr role=row>
    <th role=rowheader>...</th>
    <td role=cell>...</td>
  </tr>
</table>
```

The above markup / explicit roles are used to mitigate against inconsistencies with how browsers treat the styling of table elements, and how their role mappings can get suppressed if their CSS display changes.  

It's my opinion that the validator continue to throw warnings for the explicit use of what should be the implicit roles here, but I think there doesn't need to be an 'error' surfaced.   This would then match other rules, such as `<details>` which has an implicit `role=group`, and 'no role' is allowed, otherwise. When an author specifies a `role=group` on the element, the validator only surfaces a warning.

Does this change to the validator's output sound alright to you?  Is there anything you think I should change with the wording of these rules (and potentially others) to help facilitate this change?

Thanks
